FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/EkmLoginViewModel.kt

```diff
@@ -0,0 +1,160 @@
+/*
+ * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
+ * Contributors:
+ *  DenBond7
+ *  Ivan Pizhenko
+ */
+
+package com.flowcrypt.email.jetpack.viewmodel
+
+import android.app.Application
+import android.content.Context
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.viewModelScope
+import com.flowcrypt.email.R
+import com.flowcrypt.email.api.retrofit.ApiName
+import com.flowcrypt.email.api.retrofit.ApiRepository
+import com.flowcrypt.email.api.retrofit.FlowcryptApiRepository
+import com.flowcrypt.email.api.retrofit.request.api.DomainRulesRequest
+import com.flowcrypt.email.api.retrofit.request.api.LoginRequest
+import com.flowcrypt.email.api.retrofit.request.model.LoginModel
+import com.flowcrypt.email.api.retrofit.response.api.LoginResponse
+import com.flowcrypt.email.api.retrofit.response.base.ApiError
+import com.flowcrypt.email.api.retrofit.response.base.ApiResponse
+import com.flowcrypt.email.api.retrofit.response.base.Result
+import com.flowcrypt.email.api.retrofit.response.model.OrgRules
+import com.flowcrypt.email.util.exception.EkmNotSupportedException
+import com.flowcrypt.email.util.exception.OrgRulesCombinationNotSupportedException
+import kotlinx.coroutines.launch
+
+/**
+ * This [androidx.lifecycle.ViewModel] can be used to resolve domain rules for enterprise users.
+ *
+ * @author Denis Bondarenko
+ *         Date: 10/23/19
+ *         Time: 12:36 PM
+ *         E-mail: DenBond7@gmail.com
+ */
+class EkmLoginViewModel(application: Application) : BaseAndroidViewModel(application) {
+  private val repository: ApiRepository = FlowcryptApiRepository()
+  val ekmLiveData: MutableLiveData<Result<ApiResponse>?> = MutableLiveData()
+
+  fun fetchPrvKeys(account: String, uuid: String, tokenId: String) {
+    ekmLiveData.value = Result.loading()
+    val context: Context = getApplication()
+
+    viewModelScope.launch {
+      ekmLiveData.value = Result.loading(progressMsg = context.getString(R.string.loading))
+      try {
+        val loginResult = repository.login(
+          context,
+          LoginRequest(ApiName.POST_LOGIN, LoginModel(account, uuid), tokenId)
+        )
+
+        when (loginResult.status) {
+          Result.Status.ERROR, Result.Status.EXCEPTION -> {
+            ekmLiveData.value = loginResult
+            return@launch
+          }
+
+          Result.Status.SUCCESS -> {
+            if (loginResult.data?.isVerified == true) {
```

Comment:
My recommendation is to avoid branches that are this long - it's hard to keep track by scrolling up and down.

Typically, with code like this, there are two ways to write this:

```
if(A is good) {
  if(B is good) {
    if(C is good) {
      
    } else { // C is bad
    }
  } else { // B is bad
  
  }
} else { // A is bad

}
```

this above method is hard to keep track of. What I recommend instead is a process of elimination of bad results, and then the good result is at the end. Like this:

```
if(A is bad) {
  // deal with bad A
  return
}
if(B is bad) {
  // deal with bad B
  return
}
if(C is bad) {
  // deal with bad C
  return
}
// everything is good
```

Reply:
Got it. I will simplify it. By the way, this PR has the `draft` state.

Also, I've added reference to https://github.com/FlowCrypt/flowcrypt-android/issues/1122

Reply:
I know it's a draft :-)